source/common/listener_manager/connection_handler_impl.cc

```diff
@@ -309,6 +320,37 @@ void ConnectionHandlerImpl::setListenerRejectFraction(UnitFloat reject_fraction)
   }
 }
 
+void ConnectionHandlerImpl::enableReverseConnections(
+    Network::RevConnRegistry& reverse_conn_registry) {
+  ENVOY_LOG_MISC(debug, "Creating thread local reverse conn registry");
+  local_reverse_conn_registry_ = reverse_conn_registry.getLocalRegistry();
+  RELEASE_ASSERT(local_reverse_conn_registry_ != nullptr,
+                 "Thread local rverse conn registry should not be null.");
+}
+
+void ConnectionHandlerImpl::saveUpstreamConnection(Network::ConnectionSocketPtr&& upstream_socket,
```

Comment:
We want to store the upstream socket to a dedicated `REVESR_CLUSTER` cluster so that it can be reused when public Envoy wants to establish connections to the on-perm one. And we also wan to save upstream connection here is just for PING and other handshake processes.

The request flow is like:
1. on-prem Envoy -> public Envoy 
2. Then the public Envoy stores the downstream socket to the `REVESR_CLUSTER`, 
3. and also initialize the listener to do RPING 
4. And then the service behind the public Envoy can send request -> normal listener -> filster-chain / router -> the stored socket in REVERSE_CLUSTER.

Am I understanding correctly?

Reply:
@botengyao yes, most of the above is correct, the request flow is as follows:

Onprem -> Cloud envoy

1. Onprem initiates reverse connections; it creates http connections to cloud envoy and sends a reverse connection initiation request through it.
2. This request is intercepted by the [reverse conn http filter](https://github.com/envoyproxy/envoy/blob/af13cddb363be33c4adf913aba6bda88a32257d0/source/extensions/filters/http/reverse_conn/reverse_conn_filter.cc#L85) and cloud envoy stores the sockets with the [ReverseConnectionHandler](https://github.com/envoyproxy/envoy/blob/457e1e985a0ef2c0812f809d34aa3add2d029433/envoy/network/reverse_connection_handler.h#L40). This ReverseConnectionHandler periodically sends RPINGs over all such cached sockets.
3. On cloud envoy, a REVERSE_CONNECTION cluster type is defined and is used for all requests that need to be sent over a reverse connection. When a request arrives and this REVERSE_CONNECTION cluster is picked for the route, we interface with the ReverseConnectionHandler described above, get the cached socket and send the request over it.